isobus/include/isobus/isobus/can_parameter_group_builder.hpp

```diff
@@ -0,0 +1,503 @@
+			// of data to write.  Hence we put this version first because it covers more single
+			// byte cases.
+			std::uint8_t mask = (1 << offset) - 1;
+			buffer[byte] = (buffer[byte] & mask) | (*data << offset);
+			bits -= remaining;
+			++byte;
+
+			// -------------------------------
+			//  Stage two - copy whole bytes.
+			// -------------------------------
+
+			while (bits > 8)
+			{
+				buffer[byte] = (*data >> remaining);
+				++data;
+				buffer[byte] = (buffer[byte] & mask) | (*data << offset);
+				bits -= 8;
+				++byte;
+			}
+
+			// -------------------------------
+			//  Stage three - copy whole bytes.
+			// -------------------------------
+
+			// `bits` is number of bits left.  It may or may not span a byte boundary in the input.
+			if (bits == 0)
+			{
+				// Nothing left to copy.
+				if (writeOffset % 8 != 0)
+				{
+					// Mask the top bits to hide excess data.
+					buffer[byte] |= 255 << (writeOffset % 8);
+				}
+				return true;
+			}
+
+			// I'm sure this code can be compressed with the code above.
+			buffer[byte] = (*data >> remaining);
+			if (offset < bits)
+			{
+				// The final output spans two bytes of input.
+				++data;
+				buffer[byte] = (buffer[byte] & mask) | (*data << offset);
+			}
+			if (writeOffset % 8 != 0)
+			{
+				// Mask the top bits to hide excess data.
+				buffer[byte] |= 255 << (writeOffset % 8);
+			}
+
+			return true;
+		}
+		
+		bool read_bits(std::uint8_t * data, std::size_t bits)
+		{
+			std::size_t byte = get_read_byte_offset();
+			std::size_t input = get_read_bit_offset();
+			std::size_t remaining = 8 - input;
+			std::size_t output = 0;
+			std::size_t space = 8 - output;
+			// Mark as read, even though we actually haven't yet.
+			readOffset += bits;
+			if (readOffset > writeOffset)
+			{
+				// Trying to read too much data.
+				readOffset -= bits;
+				return false;
+			}
+			// Initialise the current destination byte.
+			*data = 0;
+			while (bits)
+			{
+				// Move in the new data.
+				*data = (*data & (255 >> space)) | ((buffer[byte] >> input) << output);
+				if (space >= bits)
+				{
+					space -= bits;
+					break;
+				}
+				if (space > remaining)
+				{
+					space -= remaining;
+					output = 8 - space;
+					remaining = 0;
+					input = 0;
+					++byte;
+				}
+				else if (remaining > space)
+				{
+					bits -= space;
+					remaining -= space;
+					input = 8 - remaining;
+					space = 8;
+					output = 0;
+					++data;
+					*data = 0;
+				}
+				else
+				{
+					bits -= space;
+					// Perfectly matched.
+					remaining = 8;
+					input = 0;
+					space = 8;
+					output = 0;
+					++data;
+					*data = 0;
+					++byte;
+				}
+			}
+			if (space)
+			{
+				// Bits left to mask.
+				*data = *data & (255 >> space);
+			}
+			return true;
+		}
+
+	public:
+		ParameterGroupBuilder() :
+		  buffer()
+		{
+		}
+
+		ParameterGroupBuilder(std::vector<std::uint8_t> const &data) :
+		  buffer()
+		{
+			buffer = data;
```

Comment:
I think this makes a copy of the data, which is not ideal. The best way to deal with this would be for the network manager to pass back a weak_ptr or shared_ptr in a CAN message and store that here, but since it's not doing that (yet), there's a lot of overhead. I also don't think we'd want to store a `std::vector<std::uint8_t> &` in the class either, since it could go out of scope in the parent scope... So I guess it's fine for the moment, but it's a fairly heavy memory price. Could refactor it if/when we improve the safety of the messages from the network manager.